Allow setting number maximum in-flight fragments for reassembly

aperezdc opened the pull request:
This adds a "fragment_cache_max" setting to the Reassembler applications (both
for IPv4 and IPv6), which limits the maximum amount of fragments pending to be
reassembled kept in the fragment cache. When the limit is reached, a random
set pending-to-reassemble fragments is evicted from the cache. The setting is
disabled by default.

The purpose of this feature is avoiding an ever-growing fragment cache which
could end up using all the available memory of the machine, and thus avoiding
potential memory exhaustion DoS situations.

aperezdc committed:
Allow setting number maximum in-flight fragments for reassembly

This adds a "fragment_cache_max" setting to the Reassembler applications (both
for IPv4 and IPv6), which limits the maximum amount of fragments pending to be
reassembled kept in the fragment cache. When the limit is reached, a random
set pending-to-reassemble fragments is evicted from the cache. The setting is
disabled by default.

The purpose of this feature is avoiding an ever-growing fragment cache which
could end up using all the available memory of the machine, and thus avoiding
potential memory exhaustion DoS situations.

kbara commented:
This needs a config file option.

aperezdc commented:
Updated to make the `fragment_cache_max` parameter mandatory, I will update in a while adding the configuration file option, and unit tests.

aperezdc committed:
Allow ipv{4,6}_fragment_cache_max options in lwaftr configuration

Modify the parser for the lwaftr configuration file so it accepts two new
options, "ipv4_fragment_cache_max" and "ipv6_fragment_cache_max", which can
be used to configure the maximum amount of pending-to-reassemble packets
the IPv4 and IPv6 Reassembler applications hold. The default value was chosen
taken into account that in Snabb Switch the packet_t structure is 16 KiB, so
each Reassembler takes at most an amount of memory around 150 MiB.

aperezdc committed:
Pass ipv{4,6}_fragment_cache_max from config file to applications

This arranges for the "ipv4_fragment_cache_max" and "ipv6_fragment_cache_max"
options from the parsed configuration files to be passed to the IPv4 and IPv6
Reassembler applications.

aperezdc committed:
Unit test for IPv4 Reassembler fragmnt cache eviction

aperezdc committed:
Unit test for IPv6 Reassembler fragmnt cache eviction

aperezdc commented:
This now has the config file option and unit tests, PTAL.

kbara commented on src/apps/lwaftr/ipv6_apps.lua:
```diff
@@ -71,15 +75,37 @@ end
+
+   if self.fragment_count >= self.fragment_cache_max then
+      local n = math.random(self.fragment_key_count)
+      for _, frags in pairs(cache) do
```
Isn't this O(N) in the number of fragment keys? .... and only called when that number is high? That's a problem.

aperezdc replied:
I have been thinking about this, and I believe to have found the best balance between simplicity and keeping all the operations O(1). Each fragment list stored in the cache would get the following additional fields:

- A `.time` field, with the time at which a packet was last added to it. It gets updated every time a fragment is added to that particular fragment list.
- `.next` and `.prev` fields, used to chain all fragment lists in a doubly-linked list. The list is kept sorted by `.time` at all times.

On top of the existing `cache` table which maps keys to fragment lists, the doubly linked list is updated whenever a new fragment arrives:

1. If the fragment list to which it belongs already exists, it is removed from the linked list (this only needs assigning `.next` and `.prev` “pointers”, it's a constant-time operation). Otherwise a new fragment list is created and addded
1. The current time is set as the `.time` of the fragment list.
1. The fragment is appended to the fragment list (also, “pointer” fiddling: constant time).
1. The fragment list is _appended_ to the fragment list.

As for trimming fragments, it can be done by traversing the linked list starting from the end. Periodic trimming of old fragments can be done in `Reassembler:push()`, as long as there is a limit in the maximum amount of elements of the linked list checked each time—to make sure that trimming does not consume more than a maximum fixed time of CPU.

aperezdc committed:
Implement utility functions for doubly linked lists

This adds new apps.lwaftr.lwutil.linked_list_*() which can be used to
manipulate circular doubly linked lists of Lua tables. The functions
add the keys "_linked_list_next" and "_linked_list_prev" to the tables
added to the list, so no additional tables are created internally for
housekeeping. The "handle" to a list is a variable with a reference to
the list's first item. Functions which modify the list may return a
changed list "handle". Quick usage example:

  local list = nil  -- Empty list
  list = linked_list_append(list, { message = "Hello" })
  list = linked_list_append(list, { message = "World" })
  while not linked_list_empty(list) do
    local item = linked_list_first(list)
    list = linked_list_remove(list, item)
    print(item.message)
  end

aperezdc committed:
Reassembly: Evict old fragments in constant time

Modifies the Reassembler apps (both IPv4 and IPv6) to keep a linked list of
fragment lists, and adds a "last seen" timestamp to them as well. The linked
list is always kept updated so the fragments most recently seen are always
at the end of the list. Evicting fragments now only needs to traverse the
first elements of the list. There are two situations in which the list of
in-flight fragments is trimmed:

 * Periodically, up 20 fragment lists are scanned, and those which have
   expired a 60s timer evicted.

 * When trying to add a fragment to the cache, if the limit of maximum
   in-flight packets would be exceeded, up to the oldest 50 fragment
   lists are evicted. This does not take into account the timer, because
   in a potential memory exhaustion situation it's better to ensure that
   the Snabb Switch process will continue running.

In the future, it might be desirable to add configurable values for the
expiration timer and the maximum amount of fragment lists scanned for
eviction.

aperezdc commented:
An additional optimization idea: before calling `self:trim()` inside `Reassembler:push()`, we can check if the timer of the _newest_ fragment list (that's the last element in the linked list) has expired. If so, we can just delete the whole list `self.fragment_list = nil`, and let the GC do its magic.

My doubt is whether that is actually such a good idea or not: suddenly the GC may find that it has a lot of work to do, and there could be some long GC pauses after clearing the list. Dunno. Opinions?

kbara commented:
That would just add the overhead of that check, in practice. There is no way a production lwaftr would see fragments that rarely.

kbara commented:
@wingo Do you vote merge, alter, or rewrite? If either of the latter, will you specifically say what you have in mind for @aperezdc or contribute a patch? If the former, let's test performance.

kbara commented:
Closing, as #400 has a superset of the reason for this change, and is merged.
